numpy/linalg/linalg.py

```diff
@@ -2346,12 +2346,13 @@ def _multi_dot_three(A, B, C):
     than `_multi_dot_matrix_chain_order`
 
     """
+    a0, a1 = A.shape
+    b1 = B.shape[1]
+    c1 = C.shape[1]
     # cost1 = cost((AB)C)
```

Comment:
Can we include the expanded formula in the comments, i.e.:

```Python
# cost1 = cost((AB)C) = a0*a1*b1 + a0*b1*c1
```

I could only make sense of your factored out code by looking at the old code, which would now be gone.

Also, it may be worth renaming your variables for clarity, `a1 -->a1b0` and `b1 --> b1c0`.